### PR TITLE
CXX-1871 add runOnRequirements and schemaVersion

### PR DESCRIPTION
**Pull request by bazile-clyde**
Most of the changes are test syncs. This PR syncs the test in "valid-pass" and adds the logic for the "runOnRequirements" and "schemaVersion" parts of the test runner. I created a new branch for the unified test format.

### CODE REVIEW

**Review by kevinAlbs — COMMENTED** [review]
I like the use of STL. Some small comments but looks great!

File: src/mongocxx/test/spec/unified_test_format.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    return std::lexicographical_compare(
+        std::begin(rng1), std::begin(rng1) + 2, std::begin(rng2), std::begin(rng2) + 2, comp);
+}
+
+bool equals_server_topology(const document::element& topologies) {
+    using bsoncxx::types::bson_value::value;
+
+    // The server's topology will not change during the test. No need to make a round-trip for every
+    // test file.
+    static std::string server_topology = test_util::get_topology();
```
[the remark this answers is not in the document]
Comment (bazile-clyde) [reply src/mongocxx/test/spec/unified_test_format.cpp]: It should! Done.

File: src/mongocxx/test/spec/unified_test_format.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    auto equals = [&](const bsoncxx::array::element& topology) {
+        return topology == value(server_topology) ||
+               (topology == value("sharded-replicaset") && server_topology == "shared");
```
[the remark this answers is not in the document]
Comment (bazile-clyde) [reply src/mongocxx/test/spec/unified_test_format.cpp]: Good catch. Yeah I ended up flippin this condition, i.e., if the test runner expects "sharded", then "sharded-replicaset" is ok. Also renamed the values to "expected" and "actual". It was getting a bit confusing.

File: src/mongocxx/test/spec/unified_test_format.cpp [hunk 3]
```diff
@@ -61,22 +58,25 @@ std::vector<int> get_version(bsoncxx::document::element doc) {
...
+template <typename Range1, typename Range2, typename Compare>
+bool is_compatible_version(Range1 range1, Range2 range2, Compare comp) {
     // only compare major and minor in version of the form "<int>.<int>.<int>", i.e., [0:2)
-    return std::lexicographical_compare(
-        std::begin(rng1), std::begin(rng1) + 2, std::begin(rng2), std::begin(rng2) + 2, comp);
+    return std::lexicographical_compare(std::begin(range1),
+                                        std::begin(range1) + 2,
+                                        std::begin(range2),
+                                        std::begin(range2) + 2,
+                                        comp);
```
Comment (kevinAlbs) [inline src/mongocxx/test/spec/unified_test_format.cpp:68]: The comparison logic may need to be tweaked slightly to account for different incompatible major versions.

If the runner only supports version `2.a.b`, then it should not consider a test of version `1.y.z` compatible.

https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#test-runner-support
> Test files are considered compatible with a test runner if their schemaVersion is less than or equal to a supported version in the test runner, given the same major version component.

As an aside: I like to think semver major version bumps as having nothing to do with the prior version. A major version bump does not guarantee any similarity. In an extreme case, a `4.0.0` C++ driver could completely remove all of the existing public API and replace it, and that would be valid by semver.

Comment (bazile-clyde) [reply src/mongocxx/test/spec/unified_test_format.cpp:68]: Done. See `is_compatible_schema_version`

**Review by kevinAlbs — APPROVED** [review]
Fantastic, LGTM!

**Pull request merged**